packages/service-library/src/servicelib/rabbitmq/_client.py

```diff
@@ -20,6 +21,63 @@
 
 _DEFAULT_PREFETCH_VALUE: Final[int] = 10
 _DEFAULT_RABBITMQ_EXECUTION_TIMEOUT_S: Final[int] = 5
+_HEADER_X_DEATH: Final[str] = "x-death"
+
+DEFAULT_UNEXPECTED_ERROR_RETRY_DELAY_S: Final[float] = 1
+DEFAULT_UNEXPECTED_ERROR_MAX_ATTEMPTS: Final[NonNegativeInt] = 15
+
+
+def _get_delayed_exchange_name(exchange_name: str) -> str:
+    return f"delayed_{exchange_name}"
+
+
+def _get_x_death_count(message: aio_pika.abc.AbstractIncomingMessage) -> int:
+    count: int = 0
+
+    x_death: list[dict[str, Any]] = message.headers.get(_HEADER_X_DEATH, [])
+    if x_death:
+        count = x_death[0]["count"]
+
+    return count
+
+
+async def _on_message(
+    message_handler: MessageHandler,
+    max_retries_upon_error: int,
+    message: aio_pika.abc.AbstractIncomingMessage,
+) -> None:
+    async with message.process(requeue=True, ignore_processed=True):
+        count = _get_x_death_count(message)
+        try:
+            with log_context(
+                _logger,
+                logging.DEBUG,
+                msg=f"Received message from {message.exchange=}, {message.routing_key=}",
+            ):
+                if not await message_handler(message.body):
+                    # NOTE: will put message to original queue(not the DLX)
+                    await message.nack()
+        except Exception:  # pylint: disable=broad-exception-caught
+            _logger.warning(
+                (
+                    "Retry [%s/%s] for handler '%s', which raised "
+                    "an unexpected error caused by message '%s'"
+                ),
+                count,
+                max_retries_upon_error,
+                message_handler,
+                message,
+            )
+
+            if count < max_retries_upon_error:
+                # NOTE: puts message to the DLX
+                await message.nack(requeue=False)
+            else:
+                _logger.error(  # noqa: TRY400
```

Comment:
I would put here exception to get more information of where is the unhandled `Exception` coming from

Reply:
Yes, good point. The tooling was also hinting at it.